node_common/upload-zip.js

```diff
@@ -0,0 +1,273 @@
+  if (!buckets) {
+    ScriptLogging.error(SHOVEL, `Utilities.getBucketAPIFromUserToken()`);
+    res.set("Connection", "close");
+    return {
+      decorator: "UPLOAD_NO_BUCKETS",
+      error: true,
+      message: `No buckets for ${user.username}.`,
+    };
+  }
+
+  let bucketSizeBytes = null;
+  try {
+    const path = await buckets.listPath(bucketKey, "/");
+    bucketSizeBytes = path.item.size;
+  } catch (e) {
+    res.set("Connection", "close");
+    return {
+      decorator: "UPLOAD_BUCKET_CHECK_FAILED",
+      error: true,
+    };
+  }
+
+  let remainingSizeBytes = NodeConstants.TEXTILE_BUCKET_LIMIT - bucketSizeBytes;
+  ScriptLogging.message(UPLOAD, `bucket size bytes : ${bucketSizeBytes}`);
+  ScriptLogging.message(UPLOAD, `remaining size bytes : ${remainingSizeBytes}`);
+
+  if (uploadSizeBytes > remainingSizeBytes) {
+    res.set("Connection", "close");
+    return {
+      decorator: "UPLOAD_NOT_ENOUGH_SPACE_REMAINS",
+      error: true,
+    };
+  }
+
+  const busboy = new BusBoyConstructor({
+    headers: req.headers,
+    highWaterMark: HIGH_WATER_MARK,
+  });
+
+  const _createStreamAndUploadToTextile = async (writableStream) => {
+    return new Promise(function (resolvePromiseFn, rejectPromiseFn) {
+      function _safeForcedSingleConcurrencyFn(actionFn, rejectFn, timeoutId) {
+        singleConcurrencyQueue.add(async function () {
+          try {
+            await actionFn();
+          } catch (e) {
+            ScriptLogging.error(SHOVEL, `${timeoutId} : queue.pause()`);
+            singleConcurrencyQueue.pause();
+
+            ScriptLogging.error(SHOVEL, `${timeoutId} : controller.abort()`);
+            controller.abort();
+
+            ScriptLogging.error(SHOVEL, `${timeoutId} : sendTextileSlackMessage()`);
+            Social.sendTextileSlackMessage({
+              file: "/node_common/upload.js",
+              user,
+              message: e.message,
+              code: e.code,
+              functionName: `${timeoutId} : _safeForcedSingleConcurrencyFn()`,
+            });
+
+            ScriptLogging.error(SHOVEL, `${timeoutId} : req.unpipe()`);
+            req.unpipe();
+
+            ScriptLogging.error(
+              SHOVEL,
+              `${timeoutId} : rejectFn() of safeForcedSingleConcurrencyFn()`
+            );
+
+            return rejectFn({
+              decorator: "UPLOAD_FAILURE",
+              error: true,
+              message: e.message,
+              id: timeoutId,
+            });
+          }
+        });
+      }
+
+      writableStream.on("file", function (fieldname, stream, filename, encoding, mime) {
+        const timeoutId = `${user.username}-${filename}`;
+
+        data = LibraryManager.createLocalDataIncomplete({
+          name: filename,
+          type: mime,
+        });
+
+        const concatStream = concat(_handleZipUpload);
+        stream.pipe(concatStream);
+
+        // (NOTE: daniel) extract and upload files
+        async function _handleZipUpload(buffer) {
```

Comment:
Is there anyway you can pull this function declaration outside of the closure here?